File: Java/pipeline/src/main/java/com/ibrahimatay/pipeline/MessagesPipeline.java
package com.ibrahimatay.pipeline;

import com.ibrahimatay.model.Messages;
import com.ibrahimatay.stage.Stage;

import java.util.ArrayList;
import java.util.List;

public class MessagesPipeline implements Pipeline<Messages>{
    final List<Stage<Messages>> stages = new ArrayList<Stage<Messages>>();
    @Override
    public void add(Stage<Messages> stage) {
        stages.add(stage);
    }

    @Override
    public void execute() {
        Messages input = null, output;

        for (Stage<Messages> stage : stages) {
            output = stage.execute(input);
            input = output;
        }
    }
}

File: Csharp/DesignPatterns.Pipeline/Pipelines/MessagesPipeline.cs
using DesignPatterns.Pipeline.Model;
using DesignPatterns.Pipeline.Stage;

namespace DesignPatterns.Pipeline.Pipelines;

public class MessagesPipeline : IPipeline<Messages>
{
    private List<IStage<Messages>> stages = new();
    
    public void Add(IStage<Messages> stage)
    {
        stages.Add(stage);
    }

    public void Execute()
    {
        Messages input = null;
        Messages output = null;

        foreach (IStage<Messages> stage in stages)
        {
            output = stage.Execute(input);
            input = output;
        }
    }
}